### PR TITLE
Port Faster R-CNN to Keras3

### PR DESCRIPTION
**Pull request by sineeli**
# What does this PR do?

Port legacy Faster R-CNN to Keras 3 

Fix minor bug in ROI Align, make sure to build the shapes for each layer components.

The present PR is modified code of [Updating FasterRCNN to use Task API](https://github.com/keras-team/keras-cv/pull/2012) contributed by @ariG23498. Please @ariG23498 any suggestions form your side are appreciated.

### CODE REVIEW

**Comment by sineeli** [comment]
Can we make last dimension of the [`roi_generator.py`](https://github.com/keras-team/keras-cv/blob/5cfe7c3d8f2d758b91bbfcedf7f0a29931236c68/keras_cv/src/layers/object_detection/roi_generator.py#L190) boxes output have fixed 4 as dimension, any suggestion on how to make it possible ?

Ref:
image

File: keras_cv/src/models/object_detection/faster_rcnn/faster_rcnn.py [hunk 1]
```diff
@@ -0,0 +1,477 @@
...
+        ) = self.rpn_labeler(
+            anchors_dict=keras.ops.concatenate(
+                tree.flatten(anchors),
+                axis=0,
+            ),
+            gt_boxes=gt_boxes,
+            gt_classes=gt_classes,
+        )
+        # 3. Computing the weights
+        rpn_box_weights /= (
```
Comment (sineeli) [inline keras_cv/src/models/object_detection/faster_rcnn/faster_rcnn.py:290]: Here for cls/box weights we need batch size and while testing I observed that it failes when given odd batch size as batch size becomes `None`. Not sure what is the causing the probelm.

The dataset constructed should be from tensor slices striagt forward `model.fit(xs, ys, epochs=1)` seems to fail as well.

image

**Comment by sineeli** [comment]
@VarunS1997 Review for any corrections are fixes required from based code.

Thanks!

**Comment by ariG23498** [comment]
Hey @sineeli this is so cool! Thanks for taking this up. It was a long time ago and unfortunately I would not be able to give you a thorough review.

What I understand should be green flags are:
- The model should follow the Functional API of Keras
- It should use the `model.fit()` API and train on PASCAL VOC

Last I worked on it I could not get every component to comply with the Functional API. If you got that part sorted, you are good to go ^_^

File: keras_cv/src/models/object_detection/faster_rcnn/faster_rcnn.py [hunk 2]
```diff
@@ -0,0 +1,477 @@
...
+        ]
+        feature_extractor = get_feature_extractor(
+            backbone, extractor_layer_names, extractor_levels
+        )
+        
+        # 2. Feature Pyramid
+        feature_pyramid = feature_pyramid or FeaturePyramid()
+        
+        # 3. Anchors
+        scales = [2**x for x in [0]]
```
Comment (VarunS1997) [inline keras_cv/src/models/object_detection/faster_rcnn/faster_rcnn.py:59]: This just evaluates to: `scales  = [1]` always
Was that intentional?

File: keras_cv/src/models/object_detection/faster_rcnn/faster_rcnn.py [hunk 3]
```diff
@@ -0,0 +1,477 @@
...
+        # 7. RCNN Head
+        rcnn_head = rcnn_head or RCNNHead(
+            num_classes,
+            name="rcnn_head"
+        )
+
+        # Begin construction of forward pass
+        image_shape =  feature_extractor.input_shape[1:]
+        if None in image_shape:
+            raise ValueError("Image shape should not have None")
```
Comment (VarunS1997) [inline keras_cv/src/models/object_detection/faster_rcnn/faster_rcnn.py:95]: Problem with this error is that it's so far removed from a user of this class, that this message, while accurate, would make no sense to them.

We should consider when/how such a None could appear in image_shape, and either instead guard against that specific situation or rephrase this error to something the user could understand.

I know this is a bit more work, but it's important I think.

File: keras_cv/src/models/object_detection/faster_rcnn/faster_rcnn.py [hunk 4]
```diff
@@ -0,0 +1,477 @@
...
+        self.bounding_box_format = bounding_box_format
+        self.anchor_generator = anchor_generator
+        self.num_classes = num_classes
+        self.rpn_labeler = label_encoder or _RpnLabelEncoder(
+            anchor_format=bounding_box_format,
+            ground_truth_box_format=bounding_box_format,
+            positive_threshold=0.7,
+            negative_threshold=0.3,
+            samples_per_image=256,
+            positive_fraction=0.5,
```
Comment (VarunS1997) [inline keras_cv/src/models/object_detection/faster_rcnn/faster_rcnn.py:172]: These args are also good candidates for input parameters of this class.

File: keras_cv/src/models/object_detection/faster_rcnn/faster_rcnn.py (hunk 1 above)
Comment (VarunS1997) [reply keras_cv/src/models/object_detection/faster_rcnn/faster_rcnn.py:290]: I would agree that we probably do need batch_size. I've never really seen it removed in operations like this before.

**Review by VarunS1997 — CHANGES_REQUESTED** [review]
Generally, good ideas here and good implementation.

We should definitely, however, implement some of the requested changes.

Also please remember to run keras_cv's linter included under the `shell/` directory.

I also will have to come back to this for architecture verification. But, if you have a working colab or similar, that would work as well.

File: keras_cv/src/models/object_detection/faster_rcnn/faster_rcnn.py (hunk 2 above)
Comment (sineeli) [reply keras_cv/src/models/object_detection/faster_rcnn/faster_rcnn.py:59]: We can make dynamically using user specific inputs, generally retinanet which is a single stage has 3 scales.

File: keras_cv/src/models/object_detection/faster_rcnn/faster_rcnn.py (hunk 4 above)
Comment (sineeli) [reply keras_cv/src/models/object_detection/faster_rcnn/faster_rcnn.py:172]: Here user can pass custom `label_encoder` or else by default these are the args we consider.

File: keras_cv/src/models/object_detection/faster_rcnn/faster_rcnn.py (hunk 3 above)
Comment (sineeli) [reply keras_cv/src/models/object_detection/faster_rcnn/faster_rcnn.py:95]: Yes, sure makes sense. Generally for us to generate anchors we need Image Shape as must input. I will figure out how can we add more context to this error and make it user friendly.

File: keras_cv/src/models/object_detection/faster_rcnn/feature_pyramid.py [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+        super().__init__(**kwargs)
+        self.conv_c2_1x1 = keras.layers.Conv2D(256, 1, 1, "same")
+        self.conv_c3_1x1 = keras.layers.Conv2D(256, 1, 1, "same")
+        self.conv_c4_1x1 = keras.layers.Conv2D(256, 1, 1, "same")
+        self.conv_c5_1x1 = keras.layers.Conv2D(256, 1, 1, "same")
+        self.conv_c2_3x3 = keras.layers.Conv2D(256, 3, 1, "same")
+        self.conv_c3_3x3 = keras.layers.Conv2D(256, 3, 1, "same")
+        self.conv_c4_3x3 = keras.layers.Conv2D(256, 3, 1, "same")
+        self.conv_c5_3x3 = keras.layers.Conv2D(256, 3, 1, "same")
+        self.conv_c6_3x3 = keras.layers.Conv2D(256, 3, 1, "same")
```
[the remark this answers is not in the document]
Comment (sineeli) [reply keras_cv/src/models/object_detection/faster_rcnn/feature_pyramid.py:36]: The main input would be levels which are from `P2` to `P5` and then based on start and end levels we can build the FPN network.

[the remark this answers is not in the document]
Comment (sineeli) [reply keras_cv/src/models/object_detection/faster_rcnn/feature_pyramid.py:36]: I observed as per retinanet and built this, also we can pass feature_pyramid as a whole network to the class. As a default we are using this. 

So to make it customizable here as well we have to add extra arguments to the class and build the network, I will try add this and we can checkout later what would be a feasible way.

File: keras_cv/src/models/object_detection/faster_rcnn/feature_pyramid.py [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+
+from keras_cv.src.api_export import keras_cv_export
+from keras_cv.src.backend import keras
+
+
+@keras_cv_export(
+    "keras_cv.models.faster_rcnn.FeaturePyramid",
+    package="keras_cv.models.faster_rcnn",
+)
+class FeaturePyramid(keras.layers.Layer):
```
[the remark this answers is not in the document]
Comment (sineeli) [reply keras_cv/src/models/object_detection/faster_rcnn/feature_pyramid.py:25]: Done

**Comment by sineeli** [comment]
There are two implementations as per torch:

`fasterrcnn_resnet50_fpn` - Present implementation similar to this config
`fasterrcnn_resnet50_fpn_v2` - Improved Configuration.

https://github.com/pytorch/vision/blob/8f73afacb81df1fcc6e2d308c5b734b126db3426/torchvision/models/detection/faster_rcnn.py#L470

File: keras_cv/src/models/object_detection/faster_rcnn/faster_rcnn.py [hunk 7]
```diff
@@ -0,0 +1,653 @@
...
+from keras_cv.src.layers.object_detection.non_max_suppression import (
+    NonMaxSuppression,
+)
+from keras_cv.src.layers.object_detection.roi_align import ROIAligner
+from keras_cv.src.layers.object_detection.roi_generator import ROIGenerator
+from keras_cv.src.layers.object_detection.roi_sampler import ROISampler
+from keras_cv.src.layers.object_detection.rpn_label_encoder import (
+    RpnLabelEncoder,
+)
+from keras_cv.src.models.object_detection.__internal__ import unpack_input
```
Comment (divyashreepathihalli) [inline keras_cv/src/models/object_detection/faster_rcnn/faster_rcnn.py:23]: lets aoid the __internal__ import for unpack_input and lets define that function inside this file instead

File: keras_cv/src/models/object_detection/faster_rcnn/faster_rcnn.py [hunk 8]
```diff
@@ -0,0 +1,653 @@
...
+        num_classes,
+        bounding_box_format,
+        anchor_generator=None,
+        feature_pyramid=None,
+        fpn_min_level=2,
+        fpn_max_level=5,
+        rpn_head=None,
+        rpn_filters=256,
+        rpn_kernel_size=3,
+        rpn_label_en_pos_th=0.7,
```
Comment (divyashreepathihalli) [inline keras_cv/src/models/object_detection/faster_rcnn/faster_rcnn.py:52]: lets abbreviate the args completely
`rpn_label_en_pos_th`-> `rpn_label_encoder_positive_threshold`. Same for all the args.

File: keras_cv/src/models/object_detection/faster_rcnn/faster_rcnn.py [hunk 9]
```diff
@@ -0,0 +1,653 @@
...
+
+
+@keras_cv_export(
+    [
+        "keras_cv.models.FasterRCNN",
+        "keras_cv.models.object_detection.FasterRCNN",
+    ]
+)
+class FasterRCNN(Task):
+    def __init__(
```
Comment (divyashreepathihalli) [inline keras_cv/src/models/object_detection/faster_rcnn/faster_rcnn.py:40]: lets add a docstring with headline, description, explain the args and add example usage.

File: keras_cv/src/models/object_detection/faster_rcnn/faster_rcnn.py [hunk 10]
```diff
@@ -0,0 +1,653 @@
...
+        rpn_label_en_neg_th=0.3,
+        rpn_label_en_samples_per_image=256,
+        rpn_label_en_pos_frac=0.5,
+        rcnn_head=None,
+        label_encoder=None,
+        prediction_decoder=None,
+        *args,
+        **kwargs,
+    ):
+        # 1. Backbone
```
Comment (divyashreepathihalli) [inline keras_cv/src/models/object_detection/faster_rcnn/faster_rcnn.py:62]: lets remove the `1.` in the comment. same for the subsequent comments.

File: keras_cv/src/models/object_detection/faster_rcnn/faster_rcnn.py [hunk 11]
```diff
@@ -0,0 +1,653 @@
...
+            raise ValueError(
+                "Expected 'classes' to be a Tensor of rank 2. "
+                f"Got y['classes'].shape={ops.shape(y['classes'])}."
+            )
+
+        gt_classes = y["classes"]
+        gt_classes = ops.expand_dims(gt_classes, axis=-1)
+
+        #######################################################################
+        # Generate  Anchors and Generate RPN Targets
```
Comment (divyashreepathihalli) [inline keras_cv/src/models/object_detection/faster_rcnn/faster_rcnn.py:320]: lets keep the comment here simple without the `#####`

File: keras_cv/src/models/object_detection/faster_rcnn/rcnn_head.py [hunk 12]
```diff
@@ -0,0 +1,97 @@
...
+from keras_cv.src.api_export import keras_cv_export
+from keras_cv.src.backend import keras
+
+
+@keras_cv_export(
+    "keras_cv.models.faster_rcnn.RCNNHead",
+    package="keras_cv.models.faster_rcnn",
+)
+class RCNNHead(keras.layers.Layer):
+    def __init__(
```
Comment (divyashreepathihalli) [inline keras_cv/src/models/object_detection/faster_rcnn/rcnn_head.py:24]: add a docstring

File: keras_cv/src/models/object_detection/faster_rcnn/rpn_head.py [hunk 13]
```diff
@@ -0,0 +1,95 @@
...
+class RPNHead(keras.layers.Layer):
+    """A Keras layer implementing the RPN architecture.
+
+    Region Proposal Networks (RPN) was first suggested in
+    [FasterRCNN](https://arxiv.org/abs/1506.01497).
+    This is an end to end trainable layer which proposes regions
+    for a detector (RCNN).
+
+    Args:
+        num_achors_per_location: The number of anchors per location.
```
Comment (divyashreepathihalli) [inline keras_cv/src/models/object_detection/faster_rcnn/rpn_head.py:18]: add args description for all the args

File: keras_cv/src/models/object_detection/faster_rcnn/feature_pyramid.py [hunk 14]
```diff
@@ -0,0 +1,244 @@
...
+
+    Example:
+    ```python
+    images = keras.layers.Input(
+        image_shape,
+        name="images",
+    )
+    extractor_levels= ["P2", "P3", "P4", "P5"]
+
+    backbone = keras_cv.models.ResNetV2Backbone.from_preset(
```
Comment (divyashreepathihalli) [inline keras_cv/src/models/object_detection/faster_rcnn/feature_pyramid.py:83]: make sure that the example in the docstring will run without errors

**Review by divyashreepathihalli — CHANGES_REQUESTED** [review]
Looking good overall. I have left a few comments. Thanks for taking this over the finish line.

**Comment by divyashreepathihalli** [comment]
@sineeli you will need to run `shell/api_gen.sh`, `shell/format.sh` and `shell/lint.sh` to resolve all the code format errors.

File: keras_cv/src/models/object_detection/faster_rcnn/feature_pyramid.py (hunk 14 above)
Comment (sineeli) [reply keras_cv/src/models/object_detection/faster_rcnn/feature_pyramid.py:79]: Done

File: keras_cv/src/models/object_detection/faster_rcnn/rpn_head.py (hunk 13 above)
Comment (sineeli) [reply keras_cv/src/models/object_detection/faster_rcnn/rpn_head.py]: Done

File: keras_cv/src/models/object_detection/faster_rcnn/faster_rcnn.py (hunk 10 above)
Comment (sineeli) [reply keras_cv/src/models/object_detection/faster_rcnn/faster_rcnn.py]: Done

File: keras_cv/src/models/object_detection/faster_rcnn/faster_rcnn.py (hunk 11 above)
Comment (sineeli) [reply keras_cv/src/models/object_detection/faster_rcnn/faster_rcnn.py:442]: Done

File: keras_cv/src/models/object_detection/faster_rcnn/faster_rcnn.py (hunk 9 above)
Comment (sineeli) [reply keras_cv/src/models/object_detection/faster_rcnn/faster_rcnn.py:146]: Done

File: keras_cv/src/models/object_detection/faster_rcnn/faster_rcnn.py (hunk 8 above)
Comment (sineeli) [reply keras_cv/src/models/object_detection/faster_rcnn/faster_rcnn.py]: Done

File: keras_cv/src/models/object_detection/faster_rcnn/faster_rcnn.py (hunk 7 above)
Comment (sineeli) [reply keras_cv/src/models/object_detection/faster_rcnn/faster_rcnn.py]: Done

File: keras_cv/src/models/object_detection/faster_rcnn/faster_rcnn_test.py [hunk 15]
```diff
@@ -0,0 +1,363 @@
...
+            tf.nest.map_structure(ops.convert_to_numpy, restored_output),
+        )
+
+    # TODO(ianstenbit): Make FasterRCNN support shapes that are not multiples
+    # of 128, perhaps by adding a flag to the anchor generator for whether to
+    # include anchors centered outside of the image. (RetinaNet does use those,
+    # while FasterRCNN doesn't). For more context on why this is the case, see
+    # https://github.com/keras-team/keras-cv/pull/1882
+    @parameterized.parameters(
+        ((2, 640, 384, 3),),
```
[the remark this answers is not in the document]
Comment (sineeli) [reply keras_cv/src/models/object_detection/faster_rcnn/faster_rcnn_test.py]: Done

File: keras_cv/src/models/object_detection/faster_rcnn/faster_rcnn_test.py [hunk 16]
```diff
@@ -0,0 +1,363 @@
...
+        self.assertEqual(len(faster_rcnn.trainable_variables), 30)
+
+    @pytest.mark.large  # Fit is slow, so mark these large.
+    @pytest.mark.skipif(not keras_3(), reason="disabling test for Keras 2")
+    def test_no_nans(self):
+        faster_rcnn = FasterRCNN(
+            num_classes=80,
+            bounding_box_format="xyxy",
+            backbone=keras_cv.models.ResNet18V2Backbone(
+                input_shape=(256, 256, 3)
```
[the remark this answers is not in the document]
Comment (sineeli) [reply keras_cv/src/models/object_detection/faster_rcnn/faster_rcnn_test.py]: Done

File: keras_cv/src/models/object_detection/faster_rcnn/faster_rcnn_test.py [hunk 17]
```diff
@@ -0,0 +1,363 @@
...
+
+    @pytest.mark.large()
+    @pytest.mark.skipif(not keras_3(), reason="disabling test for Keras 2")
+    def test_weights_contained_in_trainable_variables(self):
+        bounding_box_format = "xyxy"
+        faster_rcnn = FasterRCNN(
+            num_classes=80,
+            bounding_box_format=bounding_box_format,
+            backbone=keras_cv.models.ResNet18V2Backbone(
+                input_shape=(256, 256, 3)
```
[the remark this answers is not in the document]
Comment (sineeli) [reply keras_cv/src/models/object_detection/faster_rcnn/faster_rcnn_test.py]: Done

File: keras_cv/src/models/object_detection/faster_rcnn/faster_rcnn_test.py [hunk 18]
```diff
@@ -0,0 +1,363 @@
...
+        )
+
+    @pytest.mark.large()
+    @pytest.mark.skipif(not keras_3(), reason="disabling test for Keras 2")
+    def test_faster_rcnn_call(self):
+        faster_rcnn = FasterRCNN(
+            num_classes=80,
+            bounding_box_format="xywh",
+            backbone=keras_cv.models.ResNet18V2Backbone(
+                input_shape=(256, 256, 3)
```
[the remark this answers is not in the document]
Comment (sineeli) [reply keras_cv/src/models/object_detection/faster_rcnn/faster_rcnn_test.py]: Done

File: keras_cv/src/models/object_detection/faster_rcnn/faster_rcnn_test.py [hunk 19]
```diff
@@ -0,0 +1,363 @@
...
+
+
+class FasterRCNNTest(TestCase):
+    @pytest.mark.skipif(not keras_3(), reason="disabling test for Keras 2")
+    def test_faster_rcnn_construction(self):
+        faster_rcnn = FasterRCNN(
+            num_classes=80,
+            bounding_box_format="xyxy",
+            backbone=keras_cv.models.ResNet18V2Backbone(
+                input_shape=(256, 256, 3)
```
[the remark this answers is not in the document]
Comment (sineeli) [reply keras_cv/src/models/object_detection/faster_rcnn/faster_rcnn_test.py]: Done

**Comment by divyashreepathihalli** [comment]
Excellent work!! Thanks Siva!!

**Review by divyashreepathihalli — APPROVED** [review]
